ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1674,172 +1672,189 @@ extension SignalProducerProtocol {
+	var replayBuffers: [ObjectIdentifier: [Value]] = [:]
 
-	/// Appends a new value to the buffer, trimming it down to the given capacity
-	/// if necessary.
-	mutating func add(_ value: Value, upTo capacity: Int) {
-		precondition(capacity >= 0)
+	/// Initialize the replay state.
+	///
+	/// - parameters:
+	///   - capacity: The maximum amount of values which can be cached by the
+	///               replay state.
+	init(upTo capacity: Int) {
+		self.capacity = capacity
+	}
 
-		for buffer in replayBuffers {
-			buffer.values.append(value)
+	/// Attempt to observe the replay state.
+	///
+	/// - warning: Repeatedly observing the replay state with the same observer
+	///            should be avoided.
+	///
+	/// - parameters:
+	///   - observer: The observer to be registered.
+	///
+	/// - throws:
+	///   `ReplayError.pending(values:)` if there are still one or more values
+	///   pending for being replayed by the observer.
+	///
+	/// - returns:
+	///   The removal token of the observer, or `nil` if the replay state has
+	///   already terminated.
+	mutating func observe(_ observer: Signal<Value, Error>.Observer) -> Result<RemovalToken?, ReplayError<Value>> {
+		// Since the only use case is `replayLazily`, which always creates a unique
+		// `Observer` for every produced signal, we can use the ObjectIdentifier of
+		// the `Observer` to track them directly.
+		let id = ObjectIdentifier(observer)
+
+		func replayCompletion() -> Result<RemovalToken?, ReplayError<Value>> {
+			if let event = terminationEvent {
+				observer.action(event)
+			}
+			return .success(observers?.insert(observer))
 		}
 
-		if capacity == 0 {
-			values = []
-			return
+		if let buffer = replayBuffers[id] {
+			if buffer.isEmpty {
+				replayBuffers.removeValue(forKey: id)
+				return replayCompletion()
+			} else {
+				defer { replayBuffers[id] = [] }
+				return .failure(ReplayError(values: buffer))
+			}
+		} else {
+			if values.isEmpty {
+				return replayCompletion()
+			} else {
+				replayBuffers[id] = []
+				return .failure(ReplayError(values: values))
+			}
 		}
+	}
 
-		if capacity == 1 {
-			values = [ value ]
-			return
-		}
+	/// Enqueue the supplied event to the replay state.
+	///
+	/// - parameter:
+	///   - event: The event to be cached.
+	mutating func enqueue(_ event: Event<Value, Error>) {
+		switch event {
+		case let .next(value):
+			for key in replayBuffers.keys {
+				replayBuffers[key]!.append(value)
+			}
+
+			if capacity == 0 {
+				// `state.values` cannot be non-empty with a capacity of zero.
+				return
+			}
+
+			if capacity == 1 {
+				values = [value]
+				return
+			}
+
+			values.append(value)
 
-		values.append(value)
+			let overflow = values.count - capacity
+			if overflow > 0 {
+				values.removeSubrange(0 ..< overflow)
+			}
 
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeSubrange(0..<overflow)
+		case .failed:
+			fallthrough
+		case .completed:
+			fallthrough
+		case .interrupted:
```

Comment:
Can you list these like `case .failed, .completed, .interrupted:` instead?

(Sorry that wasn't clear before.)